File: java/109-ConvertSortedListToBinarySearchTree.java
/*

Given a singly linked list where elements are sorted in ascending order, 
convert it to a height balanced BST.

*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private ListNode curr = null;
    public TreeNode sortedListToBST(ListNode head) {
        if(head == null) return null;
        ListNode runner = head;
        int n = 0;
        while (runner != null) {
            ++n;
            runner = runner.next;
        }
        curr = head;
        return helper(n);
    }
    
    private TreeNode helper(int n) {
        if(curr == null || n == 0 ) return null;
        int m = n/2;
        TreeNode left = helper(m);
        TreeNode root = new TreeNode(curr.val);
        curr = curr.next;
        root.right = helper(n - m - 1);
        root.left = left;
        return root;
    }
}

File: c++/109-ConvertSortedListToBinarySearchTree.cpp
/*

Given a singly linked list where elements are sorted in ascending order, 
convert it to a height balanced BST.

*/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(!head) return NULL;
        if(!head->next) return new TreeNode(head->val);
        auto count = 0;
        auto *curr = head;
        while(curr) {
            ++count;
            curr = curr->next;
        }
        return helper(&head,count);
    }
 
private:
    TreeNode* helper(ListNode** head, int n) {
        if(n == 0) return NULL;
        auto m = n/2;
        auto left = helper(head, m);
        auto curr = new TreeNode((*head)->val);

        *head = (*head)->next;
        curr->left = left;
        curr->right = helper(head, n - m -1);
        return curr;
    }
};

class Solution1 {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(!head) return NULL;
        if(!head->next) return new TreeNode(head->val);
        auto count = 0;
        auto *curr = head;
        while(curr) {
            ++count;
            curr = curr->next;
        }
        return helper(&head, 0, count);
    }
 
private:
    TreeNode* helper(ListNode** head, int l, int h) {
        if(l == h) return NULL;
        auto m = l + (h-l)/2;
        auto left = helper(head, l, m);
        auto curr = new TreeNode((*head)->val);

        *head = (*head)->next;
        curr->left = left;
        curr->right = helper(head, m + 1, h);
        return curr;
    }
};
